sympy/solvers/inequalities.py

```diff
@@ -571,6 +573,47 @@ def valid(x):
             except NotImplementedError:
                 raise NotImplementedError('sorting of these roots is not supported')
 
+            #If expr contains imaginary coefficients
+            #Only real values of x for which the imaginary part is 0 are taken
+            make_real = S.Reals
+            if im(expanded_e) != S.Zero:
+                check = True
+                im_sol = FiniteSet()
+                try:
+                    a = solveset(im(expanded_e), gen, domain)
+                    if not isinstance(a, Interval):
+                        for z in a:
+                            if z not in singularities and valid(z) and z.is_real:
+                                im_sol += FiniteSet(z)
+                    else:
+                        start, end = a.inf, a.sup
+                        for z in _nsort(critical_points + FiniteSet(end)):
```

Comment:
Recommend that you `compute valid_start = valid(start)` once here after defining start and then use that variable below.

Reply:
Recommend that you compute `valid_z = valid(z)` immeadiately after starting the z loop.